`//eslint-disable-line`'s behavior is changed in 4.7.0

**Tell us about your environment**

* **ESLint Version:** 4.7.0
* **Node Version:** 8.4.0
* **npm Version:** 5.4.2

**What parser (default, Babel-ESLint, etc.) are you using?**

- default

**Please show your full configuration:**

- nothing

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->
```js
/* eslint-disable no-unused-vars */

var foo; // eslint-disable-line no-unused-vars
var bar;

/* eslint-enable no-unused-vars */
```

```
$ eslint test.js --rule no-unused-vars:error
```

**What did you expect to happen?**

No errors as same as `4.6.0`.

**What actually happened? Please include the actual, raw output from ESLint.**

```
~\dev\sandbox> eslint test.js --rule no-unused-vars:error

C:\Users\t-nagashima.MSS\dev\sandbox\test.js
  6:5  error  'bar' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

--- a/lib/util/apply-disable-directives.js
+++ b/lib/util/apply-disable-directives.js
@@ -7,6 +7,8 @@
 
 const lodash = require("lodash");
 
+const ALL_RULE_IDENTIFIER = Symbol("all");
+
 /**
  * Compares the locations of two objects in a source file
  * @param {{line: number, column: number}} itemA The first object
@@ -87,11 +89,11 @@ module.exports = options => {
 
             switch (directive.type) {
                 case "disable-line":
-                    tempDisabledByLine.add(directive.ruleId === null ? "all" : directive.ruleId);
+                    tempDisabledByLine.add(directive.ruleId === null ? ALL_RULE_IDENTIFIER : directive.ruleId);
                     break;
 
                 case "enable-line":
-                    tempDisabledByLine.delete(directive.ruleId === null ? "all" : directive.ruleId);
+                    tempDisabledByLine.delete(directive.ruleId === null ? ALL_RULE_IDENTIFIER : directive.ruleId);
                     break;
 
                 case "disable":
@@ -122,7 +124,7 @@ module.exports = options => {
 
         if (
             enabledRules.has(problem.ruleId) ||
-            (!tempDisabledByLine.has("all") && !tempDisabledByLine.has(problem.ruleId) &&
+            (!tempDisabledByLine.has(ALL_RULE_IDENTIFIER) && !tempDisabledByLine.has(problem.ruleId) &&
             !globalDisableActive && !disabledRules.has(problem.ruleId))
         ) {
             problems.push(problem);

--- a/lib/util/apply-disable-directives.js
+++ b/lib/util/apply-disable-directives.js
@@ -19,6 +19,67 @@ function compareLocations(itemA, itemB) {
 }
 
 /**
+ * This is the same as the exported function, except that it doesn't handle disable-line and disable-next-line directives.
+ * @param {Object} options options (see the exported function)
+ * @returns {Problem[]} Filtered problems (see the exported function)
+ */
+function applyDirectives(options) {
+    const problems = [];
+    let nextDirectiveIndex = 0;
+    let globalDisableActive = false;
+
+    // disabledRules is only used when there is no active global /* eslint-disable */ comment.
+    const disabledRules = new Set();
+
+    // enabledRules is only used when there is an active global /* eslint-disable */ comment.
+    const enabledRules = new Set();
+
+    for (const problem of options.problems) {
+        while (
+            nextDirectiveIndex < options.directives.length &&
+            compareLocations(options.directives[nextDirectiveIndex], problem) <= 0
+        ) {
+            const directive = options.directives[nextDirectiveIndex++];
+
+            switch (directive.type) {
+                case "disable":
+                    if (directive.ruleId === null) {
+                        globalDisableActive = true;
+                        enabledRules.clear();
+                    } else if (globalDisableActive) {
+                        enabledRules.delete(directive.ruleId);
+                    } else {
+                        disabledRules.add(directive.ruleId);
+                    }
+                    break;
+
+                case "enable":
+                    if (directive.ruleId === null) {
+                        globalDisableActive = false;
+                        disabledRules.clear();
+                    } else if (globalDisableActive) {
+                        enabledRules.add(directive.ruleId);
+                    } else {
+                        disabledRules.delete(directive.ruleId);
+                    }
+                    break;
+
+                // no default
+            }
+        }
+
+        if (
+            globalDisableActive && enabledRules.has(problem.ruleId) ||
+            !globalDisableActive && !disabledRules.has(problem.ruleId)
+        ) {
+            problems.push(problem);
+        }
+    }
+
+    return problems;
+}
+
+/**
  * Given a list of directive comments (i.e. metadata about eslint-disable and eslint-enable comments) and a list
  * of reported problems, determines which problems should be reported.
  * @param {Object} options Information about directives and problems
@@ -36,22 +97,26 @@ function compareLocations(itemA, itemB) {
  * A list of reported problems that were not disabled by the directive comments.
  */
 module.exports = options => {
-    const processedDirectives = lodash.flatMap(options.directives, directive => {
+    const blockDirectives = options.directives
+        .filter(directive => directive.type === "disable" || directive.type === "enable")
+        .sort(compareLocations);
+
+    const lineDirectives = lodash.flatMap(options.directives, directive => {
         switch (directive.type) {
             case "disable":
             case "enable":
-                return [directive];
+                return [];
 
             case "disable-line":
                 return [
-                    { type: "disable-line", line: directive.line, column: 1, ruleId: directive.ruleId },
-                    { type: "enable-line", line: directive.line + 1, column: 1, ruleId: directive.ruleId }
+                    { type: "disable", line: directive.line, column: 1, ruleId: directive.ruleId },
+                    { type: "enable", line: directive.line + 1, column: 0, ruleId: directive.ruleId }
                 ];
 
             case "disable-next-line":
                 return [
-                    { type: "disable-line", line: directive.line + 1, column: 1, ruleId: directive.ruleId },
-                    { type: "enable-line", line: directive.line + 2, column: 1, ruleId: directive.ruleId }
+                    { type: "disable", line: directive.line + 1, column: 1, ruleId: directive.ruleId },
+                    { type: "enable", line: directive.line + 2, column: 0, ruleId: directive.ruleId }
                 ];
 
             default:
@@ -59,79 +124,8 @@ module.exports = options => {
         }
     }).sort(compareLocations);
 
-    const problems = [];
-    let nextDirectiveIndex = 0;
-    let globalDisableActive = false;
+    const problemsAfterBlockDirectives = applyDirectives({ problems: options.problems, directives: blockDirectives });
+    const problemsAfterLineDirectives = applyDirectives({ problems: problemsAfterBlockDirectives, directives: lineDirectives });
 
-    // disabledRules is only used when there is no active global /* eslint-disable */ comment.
-    const disabledRules = new Set();
-
-    // enabledRules is only used when there is an active global /* eslint-disable */ comment.
-    const enabledRules = new Set();
-
-    const lineDisabledRules = new Set();
-    let globalLineDisableActive = false;
-
-    for (const problem of options.problems) {
-        while (
-            nextDirectiveIndex < processedDirectives.length &&
-            compareLocations(processedDirectives[nextDirectiveIndex], problem) <= 0
-        ) {
-            const directive = processedDirectives[nextDirectiveIndex++];
-
-            switch (directive.type) {
-                case "disable":
-                    if (directive.ruleId === null) {
-                        globalDisableActive = true;
-                        enabledRules.clear();
-                    } else if (globalDisableActive) {
-                        enabledRules.delete(directive.ruleId);
-                    } else {
-                        disabledRules.add(directive.ruleId);
-                    }
-                    break;
-
-                case "enable":
-                    if (directive.ruleId === null) {
-                        globalDisableActive = false;
-                        globalLineDisableActive = false;
-                        disabledRules.clear();
-                        lineDisabledRules.clear();
-                    } else if (globalDisableActive) {
-                        enabledRules.add(directive.ruleId);
-                    } else {
-                        disabledRules.delete(directive.ruleId);
-                        lineDisabledRules.delete(directive.ruleId);
-                    }
-                    break;
-
-                case "disable-line":
-                    if (directive.ruleId === null) {
-                        globalLineDisableActive = true;
-                        lineDisabledRules.clear();
-                    } else {
-                        lineDisabledRules.add(directive.ruleId);
-                    }
-                    break;
-
-                case "enable-line":
-                    lineDisabledRules.clear();
-                    globalLineDisableActive = false;
-                    break;
-
-                // no default
-            }
-        }
-
-        if (
-            !globalLineDisableActive && !lineDisabledRules.has(problem.ruleId) && (
-                globalDisableActive && enabledRules.has(problem.ruleId) ||
-                !globalDisableActive && !disabledRules.has(problem.ruleId)
-            )
-        ) {
-            problems.push(problem);
-        }
-    }
-
-    return problems;
+    return problemsAfterLineDirectives.sort(compareLocations);
 };

--- a/lib/util/apply-disable-directives.js
+++ b/lib/util/apply-disable-directives.js
@@ -44,14 +44,14 @@ module.exports = options => {
 
             case "disable-line":
                 return [
-                    { type: "disable", line: directive.line, column: 1, ruleId: directive.ruleId },
-                    { type: "enable", line: directive.line + 1, column: 1, ruleId: directive.ruleId }
+                    { type: "disable-line", line: directive.line, column: 1, ruleId: directive.ruleId },
+                    { type: "enable-line", line: directive.line + 1, column: 1, ruleId: directive.ruleId }
                 ];
 
             case "disable-next-line":
                 return [
-                    { type: "disable", line: directive.line + 1, column: 1, ruleId: directive.ruleId },
-                    { type: "enable", line: directive.line + 2, column: 1, ruleId: directive.ruleId }
+                    { type: "disable-line", line: directive.line + 1, column: 1, ruleId: directive.ruleId },
+                    { type: "enable-line", line: directive.line + 2, column: 1, ruleId: directive.ruleId }
                 ];
 
             default:
@@ -68,22 +68,38 @@ module.exports = options => {
 
     // enabledRules is only used when there is an active global /* eslint-disable */ comment.
     const enabledRules = new Set();
+    const lineDisabledRules = new Map();
 
     for (const problem of options.problems) {
+        let tempDisabledByLine = new Set();
+        let directive = {};
+        const lineDisabledRule = lineDisabledRules.get(problem.line);
+
+        if (lineDisabledRule && problem.column - lineDisabledRule.column >= 0) {
+            tempDisabledByLine = lineDisabledRule.disabledRules;
+        }
+
         while (
             nextDirectiveIndex < processedDirectives.length &&
             compareLocations(processedDirectives[nextDirectiveIndex], problem) <= 0
         ) {
-            const directive = processedDirectives[nextDirectiveIndex++];
+            directive = processedDirectives[nextDirectiveIndex++];
 
             switch (directive.type) {
+                case "disable-line":
+                    tempDisabledByLine.add(directive.ruleId === null ? "all" : directive.ruleId);
+                    break;
+
+                case "enable-line":
+                    tempDisabledByLine.delete(directive.ruleId === null ? "all" : directive.ruleId);
+                    break;
+
                 case "disable":
                     if (directive.ruleId === null) {
                         globalDisableActive = true;
                         enabledRules.clear();
-                    } else if (globalDisableActive) {
-                        enabledRules.delete(directive.ruleId);
                     } else {
+                        enabledRules.delete(directive.ruleId);
                         disabledRules.add(directive.ruleId);
                     }
                     break;
@@ -92,6 +108,7 @@ module.exports = options => {
                     if (directive.ruleId === null) {
                         globalDisableActive = false;
                         disabledRules.clear();
+                        tempDisabledByLine.clear();
                     } else if (globalDisableActive) {
                         enabledRules.add(directive.ruleId);
                     } else {
@@ -104,11 +121,17 @@ module.exports = options => {
         }
 
         if (
-            globalDisableActive && enabledRules.has(problem.ruleId) ||
-            !globalDisableActive && !disabledRules.has(problem.ruleId)
+            enabledRules.has(problem.ruleId) ||
+            (!tempDisabledByLine.has("all") && !tempDisabledByLine.has(problem.ruleId) &&
+            !globalDisableActive && !disabledRules.has(problem.ruleId))
         ) {
             problems.push(problem);
         }
+
+        lineDisabledRules.set(problem.line, {
+            column: directive.column || 0,
+            disabledRules: tempDisabledByLine
+        });
     }
 
     return problems;

--- a/lib/util/apply-disable-directives.js
+++ b/lib/util/apply-disable-directives.js
@@ -7,8 +7,6 @@
 
 const lodash = require("lodash");
 
-const ALL_RULE_IDENTIFIER = Symbol("all");
-
 /**
  * Compares the locations of two objects in a source file
  * @param {{line: number, column: number}} itemA The first object
@@ -70,38 +68,25 @@ module.exports = options => {
 
     // enabledRules is only used when there is an active global /* eslint-disable */ comment.
     const enabledRules = new Set();
-    const lineDisabledRules = new Map();
+
+    const lineDisabledRules = new Set();
+    let globalLineDisableActive = false;
 
     for (const problem of options.problems) {
-        let tempDisabledByLine = new Set();
-        let directive = {};
-        const lineDisabledRule = lineDisabledRules.get(problem.line);
-
-        if (lineDisabledRule && problem.column - lineDisabledRule.column >= 0) {
-            tempDisabledByLine = lineDisabledRule.disabledRules;
-        }
-
         while (
             nextDirectiveIndex < processedDirectives.length &&
             compareLocations(processedDirectives[nextDirectiveIndex], problem) <= 0
         ) {
-            directive = processedDirectives[nextDirectiveIndex++];
+            const directive = processedDirectives[nextDirectiveIndex++];
 
             switch (directive.type) {
-                case "disable-line":
-                    tempDisabledByLine.add(directive.ruleId === null ? ALL_RULE_IDENTIFIER : directive.ruleId);
-                    break;
-
-                case "enable-line":
-                    tempDisabledByLine.delete(directive.ruleId === null ? ALL_RULE_IDENTIFIER : directive.ruleId);
-                    break;
-
                 case "disable":
                     if (directive.ruleId === null) {
                         globalDisableActive = true;
                         enabledRules.clear();
-                    } else {
+                    } else if (globalDisableActive) {
                         enabledRules.delete(directive.ruleId);
+                    } else {
                         disabledRules.add(directive.ruleId);
                     }
                     break;
@@ -109,13 +94,29 @@ module.exports = options => {
                 case "enable":
                     if (directive.ruleId === null) {
                         globalDisableActive = false;
+                        globalLineDisableActive = false;
                         disabledRules.clear();
-                        tempDisabledByLine.clear();
+                        lineDisabledRules.clear();
                     } else if (globalDisableActive) {
                         enabledRules.add(directive.ruleId);
                     } else {
                         disabledRules.delete(directive.ruleId);
+                        lineDisabledRules.delete(directive.ruleId);
                     }
+                    break;
+
+                case "disable-line":
+                    if (directive.ruleId === null) {
+                        globalLineDisableActive = true;
+                        lineDisabledRules.clear();
+                    } else {
+                        lineDisabledRules.add(directive.ruleId);
+                    }
+                    break;
+
+                case "enable-line":
+                    lineDisabledRules.clear();
+                    globalLineDisableActive = false;
                     break;
 
                 // no default
@@ -123,17 +124,13 @@ module.exports = options => {
         }
 
         if (
-            enabledRules.has(problem.ruleId) ||
-            (!tempDisabledByLine.has(ALL_RULE_IDENTIFIER) && !tempDisabledByLine.has(problem.ruleId) &&
-            !globalDisableActive && !disabledRules.has(problem.ruleId))
+            !globalLineDisableActive && !lineDisabledRules.has(problem.ruleId) && (
+                globalDisableActive && enabledRules.has(problem.ruleId) ||
+                !globalDisableActive && !disabledRules.has(problem.ruleId)
+            )
         ) {
             problems.push(problem);
         }
-
-        lineDisabledRules.set(problem.line, {
-            column: directive.column || 0,
-            disabledRules: tempDisabledByLine
-        });
     }
 
     return problems;

--- a/tests/lib/linter.js
+++ b/tests/lib/linter.js
@@ -2072,6 +2072,19 @@ describe("Linter", () => {
             assert.equal(messages.length, 0);
         });
 
+        it("should report no violation", () => {
+            const code = [
+                "/* eslint-disable quotes */",
+                "console.log(\"foo\");",
+                "/* eslint-enable quotes */"
+            ].join("\n");
+            const config = { rules: { quotes: 2 } };
+
+            const messages = linter.verify(code, config, filename);
+
+            assert.equal(messages.length, 0);
+        });
+
         it("should report a violation", () => {
             const code = [
                 "/*eslint-disable no-alert, no-console */",

--- a/tests/lib/linter.js
+++ b/tests/lib/linter.js
@@ -1833,7 +1833,7 @@ describe("Linter", () => {
             it("should not report a violation", () => {
                 const code = [
                     "alert('test'); // eslint-disable-line no-alert",
-                    "console('test'); // eslint-disable-line no-console"
+                    "console.log('test'); // eslint-disable-line no-console"
                 ].join("\n");
                 const config = {
                     rules: {
@@ -1850,7 +1850,7 @@ describe("Linter", () => {
             it("should not report a violation", () => {
                 const code = [
                     "alert('test') // eslint-disable-line no-alert, quotes, semi",
-                    "console('test'); // eslint-disable-line"
+                    "console.log('test'); // eslint-disable-line"
                 ].join("\n");
                 const config = {
                     rules: {
@@ -2041,6 +2041,35 @@ describe("Linter", () => {
             assert.equal(messages.length, 1);
 
             assert.equal(messages[0].ruleId, "no-console");
+        });
+
+        it("should report no violation", () => {
+            const code = [
+                "/*eslint-disable no-unused-vars */",
+                "var foo; // eslint-disable-line no-unused-vars",
+                "var bar;",
+                "/* eslint-enable no-unused-vars */" // here
+            ].join("\n");
+            const config = { rules: { "no-unused-vars": 2 } };
+
+            const messages = linter.verify(code, config, filename);
+
+            assert.equal(messages.length, 0);
+        });
+
+        it("should report no violation", () => {
+            const code = [
+                "var foo1; // eslint-disable-line no-unused-vars",
+                "var foo2; // eslint-disable-line no-unused-vars",
+                "var foo3; // eslint-disable-line no-unused-vars",
+                "var foo4; // eslint-disable-line no-unused-vars",
+                "var foo5; // eslint-disable-line no-unused-vars"
+            ].join("\n");
+            const config = { rules: { "no-unused-vars": 2 } };
+
+            const messages = linter.verify(code, config, filename);
+
+            assert.equal(messages.length, 0);
         });
 
         it("should report a violation", () => {

--- a/tests/lib/util/apply-disable-directives.js
+++ b/tests/lib/util/apply-disable-directives.js
@@ -143,6 +143,34 @@ describe("apply-disable-directives", () => {
             );
         });
 
+        it("filter out problems if disable all then enable foo and then disable foo", () => {
+            assert.deepEqual(
+                applyDisableDirectives({
+                    directives: [
+                        { type: "disable", line: 1, column: 1, ruleId: null },
+                        { type: "enable", line: 1, column: 5, ruleId: "foo" },
+                        { type: "disable", line: 2, column: 1, ruleId: "foo" }
+                    ],
+                    problems: [{ line: 3, column: 3, ruleId: "foo" }]
+                }),
+                []
+            );
+        });
+
+        it("filter out problems if disable all then enable foo and then disable all", () => {
+            assert.deepEqual(
+                applyDisableDirectives({
+                    directives: [
+                        { type: "disable", line: 1, column: 1, ruleId: null },
+                        { type: "enable", line: 1, column: 5, ruleId: "foo" },
+                        { type: "disable", line: 2, column: 1, ruleId: null }
+                    ],
+                    problems: [{ line: 3, column: 3, ruleId: "foo" }]
+                }),
+                []
+            );
+        });
+
         it("keeps problems before the eslint-enable comment if there is no corresponding disable comment", () => {
             assert.deepEqual(
                 applyDisableDirectives({

--- a/tests/lib/util/apply-disable-directives.js
+++ b/tests/lib/util/apply-disable-directives.js
@@ -326,6 +326,23 @@ describe("apply-disable-directives", () => {
                 []
             );
         });
+
+        it("handles consecutive comments appropriately", () => {
+            assert.deepEqual(
+                applyDisableDirectives({
+                    directives: [
+                        { type: "disable-line", line: 1, column: 5, ruleId: "foo" },
+                        { type: "disable-line", line: 2, column: 5, ruleId: "foo" },
+                        { type: "disable-line", line: 3, column: 5, ruleId: "foo" },
+                        { type: "disable-line", line: 4, column: 5, ruleId: "foo" },
+                        { type: "disable-line", line: 5, column: 5, ruleId: "foo" },
+                        { type: "disable-line", line: 6, column: 5, ruleId: "foo" }
+                    ],
+                    problems: [{ line: 2, column: 1, ruleId: "foo" }]
+                }),
+                []
+            );
+        });
     });
 
     describe("eslint-disable-next-line comments without rules", () => {
@@ -369,19 +386,6 @@ describe("apply-disable-directives", () => {
                     problems: [{ line: 2, column: 2, ruleId: "foo" }]
                 }),
                 []
-            );
-        });
-
-        it("keeps problems on the next line if there is an eslint-enable comment before the problem on the next line", () => {
-            assert.deepEqual(
-                applyDisableDirectives({
-                    directives: [
-                        { type: "disable-next-line", line: 1, column: 1, ruleId: null },
-                        { type: "enable", line: 2, column: 1, ruleId: null }
-                    ],
-                    problems: [{ line: 2, column: 2, ruleId: "foo" }]
-                }),
-                [{ line: 2, column: 2, ruleId: "foo" }]
             );
         });
     });

